libcudacxx/include/cuda/__bit/bit_reverse.h

```diff
@@ -0,0 +1,167 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA___BIT_BIT_REVERSE_H
+#define _CUDA___BIT_BIT_REVERSE_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__type_traits/is_constant_evaluated.h>
+#include <cuda/std/__type_traits/is_unsigned_integer.h>
+#include <cuda/std/cstdint>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+#if _CCCL_COMPILER(CLANG)
+
+template <typename _Tp>
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp __bit_reverse_clang(_Tp __value) noexcept
+{
+#  if _CCCL_HAS_INT128()
+  if constexpr (sizeof(_Tp) == sizeof(__uint128_t))
+  {
+    auto __high = static_cast<__uint128_t>(__builtin_bitreverse64(static_cast<uint64_t>(__value))) << 64;
+    auto __low  = static_cast<__uint128_t>(__builtin_bitreverse64(static_cast<uint64_t>(__value >> 64)));
+    return __high | __low;
+  }
+#  endif // _CCCL_HAS_INT128()
+  if constexpr (sizeof(_Tp) == sizeof(uint64_t))
+  {
+    return __builtin_bitreverse64(__value);
```

Comment:
We conventionally always use our internal macros for compiler builtins. That way we can be safe that the builtin exists and we can also ensure that if compiler support improves we do not need any code changes